neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+
+### Curve points encoding
+
+#### Bool as output
+
+The bool is encoded as a little-endian `u64` type in Rust. The `true` value is encoded as `1` and `false` is encoded as `0`. All other values of `u64` are not allowed and should be interpreted as incorrect. Encoding is the same as for BN254 implementation in nearcore[^50].
+
+#### Sign
+
+The sign of the point on the elliptic curve is encoded as `u8` type in Rust with two possible values: `0` and `1`. `0` for positive sign, and `1` for negative sign. All other value of `u8` is not allowed and should be interpreted as incorrect.
+
+#### Scalar
+
+The scalar value is encoded as a big-endian `[u8;32]`. All possible bytes combination is allowed. Encoding is similar with BN254 implementation in nearcore[^50], but `big-endian` encoding is used instead of `little-endian` as in EIP-2537[^15].
+
+#### Fields elements Fp
+
+The value from $F_p$ is encoded as a big-endian `[u8; 48]`. Only values less than `p` are allowed. If the value is equal to or bigger than `p` the error should be returned.
+
+The rule of encoding is consistent with zkcrypto[^53], with implementation in milagro lib[^29].
+
+#### Extension fields elements Fp2
+
+The $q \in F_{p^{2}}$ could be written as $q = c_0 + c_1 v$, where $c_0, c_1 \in F_p$ The element from $F_{p^2}$ encoded in `[u8; 96]` as a bytes’ concatenation of $c_1$ and $c_0$. The $c_1$ and $c_0$ are encoded by the rule described in the previous section.
+
+$q \in F_{p^2}$, $q = c_0 + c_1 v$ encoded as `[u8; 96]`:
+
+- $c_1 \in F_p$ `[u8; 48]`
+- $c_0 \in F_p$ `[u8; 48]`
+
+The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].
+
+#### Uncompressed points on curve E(Fp)
+
+The points on the curve represent by affine coordinates: `(x: Fp, y: Fp)`. The elements from $E(F_p)$ encoded in `[u8; 96]` as bytes’ concatenation of `x` and `y` point coordinates, where $x, y \in F_p$. The $x, y$ are encoded according to the rules described in the section “Fields elements $F_p$” .
+
+$E(F_p)$  is encoded as `[u8; 96]`:
+
+- $x \in F_p$ `[u8; 48]`
+- $y \in F_p$ `[u8; 48]`
+
+*The second bit* is used to mark a point on infinity, if the second bit is set to 1 — it is an infinity point.
+
+Encoding point on infinity:
+
+```bash
+let x: [u8; 96] = [0; 96];
+x[0] = x[0] | 0x40;
+```
+
+The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].
+
+#### Compressed points on curve E(Fp)
+
+The points on the curve represent by affine coordinates: `(x: Fp, y: Fp)`. The elements from $E(F_p)$ in compressed form are encoded in `[u8; 48]` as *big-endian* encoded $x \in F_p$. The $y$  coordinate can be detected by the formula: $y = \pm \sqrt{x^3 + 4}$.
+
+*The first bit* should be set as 1. This bit indicates that point is encoded in compressed form.
```

Comment:
Similar to the above, this makes it consistent with the existing standards. It is debatable whether it makes sense to use a nonstandard but "better" format (i.e. little endian, without extra bits, etc.), this NEP proposes to stick to the standard.

Reply:
@Ekleog-NEAR 

I want to follow some existing encoding format and not develop my own. There are a few different candidates for that. For example, first I thought to encode it in the same way as in EIP-2537. 

But in EIP-2537 only uncompressed points are used. In this NEP we have functions that take the compressed points (decompress functions). As a result, we should have an encoding format for these points as well. 

In the end, I decided to follow ZCash encoding format: https://github.com/supranational/blst#serialization-format

This format is used in different BLS libraries(blst, milagro and so on) and it is also used for encoding public keys, and signatures in BLS-signatures in Ethereum. 

And I prefer to follow exactly this format even if it has some redundancy. I use this format to be compatible with different libraries and applications.  

Also, if we follow this format it will reduce the bugs' probability in implementation. Because in that case, we need not to re-encode data. 

Most probably this bit will be set anyway because the data were serialized by using some BLS library and if the bit is not set, more likely that the data is incorrect.